### DISCUSSION TITLE
What's the advantage of unit tests over functional tests?

### DISCUSSION THREAD

**Discussion by ThomasLandauer** [Codeception/Codeception#6096] [General]
While trying to solve https://github.com/Codeception/module-doctrine2/issues/38#issuecomment-766090088, I asked myself: Why do I write some tests into unit suite at all? What's the advantage of the unit suite over the functional suite? (The *dis*advantage is that you have no actors and therefore cannot do things like `amOnPage()`, but that's not the question.)

In other words: If you have a functional suite anyway, wouldn't it be easier to just enable the Asserts module (which you probably have anyway), rather than creating another suite for unit tests?

**Comment by Naktibalda** [comment]
Unit tests is a different type of tests - they test specific isolated functionality. Functional tests run entire application to test if it all works together.
Unit tests are usually written by developers as part of TDD process and they are a few orders of magnitude faster than functional tests.
I have a feeling that you are a QA person (you test a lot more than I do) and don't write any unit tests.

It is good for organizational reasons to separate tests that test specific isolated functionality from big picture tests.
And it is good for performance reasons, because unnecessary modules slow down tests signifficantly.

**Comment by ThomasLandauer** [comment]
> I have a feeling that you are a QA person

I'm an all-in-one person :-) Manager, coder, tester, coffee maker... And I do write real unit tests, but I usually don't do TDD.

Anyway, the idea I'm having: Those *three* pre-generated suites make it a little confusing for beginners. So I'm thinking about telling them something like: If you have some functional tests and only a handful unit tests, it's OK to write them in functional suite.

What you're saying about organization and process is certainly true for some users (i.e. large companies), but not for all - so ultimately they have to decide themselves how to organize their tests.

So what I wanted to ask is: What's the difference behind the scenes between `$I->assertSame(...)` in a functional test and `$this->assertSame(...)` in a unit test? In the end, it's the Asserts module, isn't it?

**Reply by Naktibalda** [reply]
> What's the difference behind the scenes between $I->assertSame(...) in a functional test and $this->assertSame(...) in a unit test?

I told you somewhere else that unit tests are PHPUnit tests, so `$this->assertSame()` is calling PHPUnit\Framework\TestCase::assertSame directly.

> If you have some functional tests and only a handful unit tests, it's OK to write them in functional suite.

If they know what unit tests are, they should be fine with writing them in different directory.
They don't need bad advice how to organize their tests.

3 pre-generated directories is not the only option since Codeception 3.0
Codeception has [Installation Templates](https://codeception.com/docs/08-Customization#Installation-Templates) that allow creating only one suite, but it seems that they are underdocumented and not mentioned in Getting Started page.

**Reply by ThomasLandauer** [reply]
> so $this->assertSame() is calling PHPUnit\Framework\TestCase::assertSame directly.

Whereas in functional tests, `$I->assertSame()` is handled by Asserts module?! Which passes it to PHPUnit or not?

**Reply by Naktibalda** [reply]
yes, Asserts module calls the same PHPUnit method.

**Comment by TavoNiievez** [comment]
An interesting topic.

I usually follow the testing pyramid.

![imagen](https://user-images.githubusercontent.com/64917965/106219836-35350800-61a8-11eb-9a4d-be342a50ee9b.png)

I am more concerned with writing unit tests than integration or acceptance tests.

Although BDD (Gherkin tests) are great for sharing tests with non-technical people they are almost always perceived as additional (ergo, unnecessary) steps.

**Reply by c33s** [reply]
functional testing == comonent testing in this pic?

BDD can really be helpful to let the product owner define the/some tests.

**Reply by TavoNiievez** [reply]
> functional testing == comonent testing in this pic?

Yes, Functional testing = Framework testing. Framework = Set of components.
